### PR TITLE
BL-790-basic-data-presentations

### PR DESCRIPTION
**Pull request by mack-gallagher**
## Description

Wire out `graphs` route [ previously `heatmaps` ] to show users a visualization of asylum case outcome data specific to their query.

### *Note on PR size*

**Most** of the nominal ~5,200 lines of code that were added as part of this PR, can actually be credited to how we added a completely new `test_data` file [ under `src/data` ], consisting of ~3,000 lines of raw JSON, for mocking API calls. 

So in terms of actual code, we **added** approximately 2,200 lines, but **removed** 2,800! 

### Routing / UX

We use routes and React Router to keep track of which view the user is currently viewing. [ All of the functions for this are currently stored in different components in the `DataVisualizations` directory. ] This way the user can switch between multiple stored views of queried data.

### State

We store Plotly- and table- ready data **for each individual view**, of which there are potentially 23 total - 10 offices * 2 office-specific views, + 3 general views - in Redux. This way the user can make a query for, say, grant rates over time [ Time Series view ] for the ZNK office from 2016 - 2021, then tab over to view a heatmap showing grant rates per year for each office, and then come back to the ZNK office, and return seamlessly to a visualization of data from the same query they were looking at before.

### API calls

We make API calls from an antd form within the `YearLimitsSelect` component. A function listening on the `Update Query` button: [ 1 ] calls a function to fetch our data from axios [ currently "fetching" mock data ], then [ 2 ] filters that data using the rawApiDataToPlotlyReadyInfo function we have in the project `utils` directory, **then** [ 3 ] dispatches an action to set the proper Redux state according to [ a ] the Plot.ly-prepared info it received, AND [ b ] additional arguments it was passed about **which view** the state should be updated for.

### Plot.ly / Tables [ Data Display ]

We use plotly.js-react to render tables according to the data received. We use 3 types of Plot.ly graphs - "scatter", for the Time Series views, "heatmap", for the Offices Heat Map view, and "chloropleth" for the Citizenship Map views.

We also display the summary data in a tabular format. Currently, this is in the form of a simple custom Table component I wrote due to since-discarded plans to render sub-tables that would not have worked with Ant Design, but it can easily be rewritten to use the standard Ant Design Table component. The Table view consumes **differently formatted data** from the Plot.ly graphs, which means that the rawApiDataToPlotlyReadyInfo function passes data pre-formatted in **2 distinct ways** to Redux, and then Redux stores those 2 forms of pre-formatted data, for each user view.

#### Video Link

[ Loom 1 [ UX ] ](https://www.loom.com/share/f0c09fea1c0448a89e8780dc59675d03)
[ Loom 2 [ code I ] ](https://www.loom.com/share/13e1c5a7fb0e49c98d77fe122cd3b2c0)
[ Loom 3 [ code II ] ](https://www.loom.com/share/fb102257b0a54ac1a0551188c9ab22c9)

#### Jira Link

https://bloomtechlabs.atlassian.net/browse/BL-790?atlOrigin=eyJpIjoiNDRkNDg2ZjQ3ZjY0NDc2Y2JkZWU5ZmFhYzk1ZmJlNjYiLCJwIjoiaiJ9

## Type of change

___Please delete options that are not relevant.___

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project 
- [x] I have performed a self-review of my own code
- [x] I have removed unnecessary comments/console logs from my code
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings
- [x] No duplicate code left within changed files
- [x] Size of pull request kept to a minimum ( [sic] -Mack )
- [x] Pull request description clearly describes changes made & motivations for said changes

### CODE REVIEW

File: src/components/pages/DataVisualizations/YearLimitsSelect.jsx
Comment (ryan-hamblin) [inline src/components/pages/DataVisualizations/YearLimitsSelect.jsx:98]: LOVE this design pattern. Functional programming is a great way to organize and think about the work that these components need in order to display the data that the user wants to see. 

There are a few ways we can optimize some of this but for now this shows an extreme level of engineering effort and ability.

Comment (mack-gallagher) [reply src/components/pages/DataVisualizations/YearLimitsSelect.jsx:98]: Thank you!

One question: Ash had mentioned checking out a branch with the "feature/BL-790-fe" prefix, and replacing this draft PR with one from the properly prefixed branch. Should I go ahead with that or leave this [ branch name / PR ] for consistency?

**Review by EstrellaDionis — APPROVED** [review]
very nice work. The code itself looks great and the UX looks very easy to use. Great job going in depth in your video explanations as well!

**Review by XavierTwitty — DISMISSED** [review]
Good job UX looks great

**Comment by Austin-T-Johnson** [comment]
Outstanding work on the implementation of the visualizations here! This was a big PR and you did a great job walking us through the entire codebase with your loom videos.

File: .gitignore
Comment (ashtilawat23) [inline .gitignore:32]: I don't think we should exclude the prettierignore, could you actually put it back? What was the reason you were thinking we had to do this?

File: src/components/Layout/Footer.jsx
Comment (ashtilawat23) [inline src/components/Layout/Footer.jsx:45]: Would it be possible to map through an object for this data? Might reduce the lines of codes significantly.

File: src/components/pages/DataVisualizations/GraphWrapper.jsx
Comment (ashtilawat23) [inline src/components/pages/DataVisualizations/GraphWrapper.jsx:86]: Could we move the styles to a file designated for this file?

File: src/components/pages/DataVisualizations/Graphs/CitizenshipMapAll.jsx
Comment (ashtilawat23) [inline src/components/pages/DataVisualizations/Graphs/CitizenshipMapAll.jsx:52]: When you use `.hasOwnProperty` how do you expect to unit this? With proptypes, unit testing is pretty straight forward. Why is `.hasOwnProperty` better than that?

File: src/components/pages/DataVisualizations/Graphs/CitizenshipMapSingleOffice.jsx
Comment (ashtilawat23) [inline src/components/pages/DataVisualizations/Graphs/CitizenshipMapSingleOffice.jsx:8]: Please remove `primary_accent_color` and `secondary_accent_color` from this file if it is not being used.

File: src/components/pages/DataVisualizations/Graphs/OfficeHeatMap.jsx
Comment (ashtilawat23) [inline src/components/pages/DataVisualizations/Graphs/OfficeHeatMap.jsx:26]: This file is well done!

File: src/components/pages/DataVisualizations/Graphs/TimeSeriesAll.jsx
Comment (ashtilawat23) [inline src/components/pages/DataVisualizations/Graphs/TimeSeriesAll.jsx:43]: This file is well done!

File: src/components/pages/DataVisualizations/YearLimitsSlider/index.jsx
Comment (ashtilawat23) [inline src/components/pages/DataVisualizations/YearLimitsSlider/index.jsx:3]: Just want to make sure every file deleted was intended to be deleted and we did not just want to maintain the version on main. Deleting a file is different than not including it in your PR. Just want to make sure, as these will be deleted completely from `main`.

**Review by ashtilawat23 — CHANGES_REQUESTED** [review]
Amazing work! This was much needed. 

Some questions about: 
- use of inline styles vs. designated files
- using `.hasPropertyType` in unit tests
- adding a `.map` where UI is repeated

File: .gitignore
Comment (mack-gallagher) [reply .gitignore:32]: There actually wasn't one in the repo - I put one there because I was getting annoyed at the linter changes. I'm not sure if it even did anything. I can remove it from the .gitignore and make that commit!

Comment (mack-gallagher) [reply .gitignore]: It should be done!

File: src/components/pages/DataVisualizations/Graphs/CitizenshipMapSingleOffice.jsx
Comment (mack-gallagher) [reply src/components/pages/DataVisualizations/Graphs/CitizenshipMapSingleOffice.jsx:8]: Done, thank you!

File: src/components/pages/DataVisualizations/GraphWrapper.jsx
Comment (mack-gallagher) [reply src/components/pages/DataVisualizations/GraphWrapper.jsx:86]: I would be happy to extract styles and organize them in external files; it feels like something that would maybe be good for another ticket?

File: src/components/Layout/Footer.jsx
Comment (mack-gallagher) [reply src/components/Layout/Footer.jsx:45]: Sorry, I'm having trouble picturing what you're imagining; could you elaborate a little?

Comment (mack-gallagher) [reply src/components/Layout/Footer.jsx:45]: Oh, never mind - now I see the repeated UI you are pointing at! Yes, we can for sure map to generate the footer components.

Comment (mack-gallagher) [reply src/components/Layout/Footer.jsx]: It is done!

File: src/components/pages/DataVisualizations/Graphs/CitizenshipMapAll.jsx
Comment (Epeng1994) [reply src/components/pages/DataVisualizations/Graphs/CitizenshipMapAll.jsx]: Hi Ash, I rewrote the .hasOwnProperty lines to useState and useEffect instead. Originally the .hasOwnProperty was used in a ternary expression to check if it exists within the props object coming from Redux Store however, it is now using useState/effect. Each graph component will have the row data set to an empty array on render if there's no associated rowData key to prevent page from crashing due to undefined data and will be updated once user submits a query.

File: .gitignore
Comment (ashtilawat23) [reply .gitignore]: Thanks for adding it!

File: src/components/Layout/Footer.jsx
Comment (ashtilawat23) [reply src/components/Layout/Footer.jsx]: Thank you Mack!

File: src/components/pages/DataVisualizations/GraphWrapper.jsx
Comment (ashtilawat23) [reply src/components/pages/DataVisualizations/GraphWrapper.jsx:125]: Let's make an GitHub issue for this. And then with John's approval, we can transfer it into a ticket.

File: src/components/pages/DataVisualizations/Graphs/CitizenshipMapAll.jsx
Comment (ashtilawat23) [reply src/components/pages/DataVisualizations/Graphs/CitizenshipMapAll.jsx]: Makes sense! Thank you for clarifying your thought process.

File: src/components/pages/DataVisualizations/Graphs/CitizenshipMapSingleOffice.jsx
Comment (ashtilawat23) [reply src/components/pages/DataVisualizations/Graphs/CitizenshipMapSingleOffice.jsx:8]: Amazing thanks!

File: src/components/pages/DataVisualizations/YearLimitsSlider/index.jsx
Comment (ashtilawat23) [reply src/components/pages/DataVisualizations/YearLimitsSlider/index.jsx:3]: Thank you for checking twice!

File: src/components/pages/DataVisualizations/YearLimitsSelect.jsx
Comment (ashtilawat23) [reply src/components/pages/DataVisualizations/YearLimitsSelect.jsx:99]: For the Jira ticket connection, it's important to have the Jira issue key at the beginning of the PR. I would change the branch name over to the properly prefixed as you were saying.

**Comment by ashtilawat23** [comment]
Next steps: Going to watch the Loom videos, wait for the branch name to be changed to the correct prefixed one, and wait for some reviews from teammates since the old reviews have gone stale.

**Review by ashtilawat23 — APPROVED** [review]

**Pull request merged**
